lib/services/local/generic/generic_wrapper_test.go

```diff
@@ -0,0 +1,225 @@
+	require.NoError(t, err)
+	require.Equal(t, r1, r1clone)
+
+	// Initially we expect no resources.
+	out, nextToken, err := service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Empty(t, out)
+
+	// Create both resources.
+	r1, err = service.CreateResource(ctx, r1)
+	require.NoError(t, err)
+	r2, err = service.CreateResource(ctx, r2)
+	require.NoError(t, err)
+
+	require.NotEmpty(t, r1.GetMetadata().GetRevision())
+	require.NotEmpty(t, r2.GetMetadata().GetRevision())
+	require.NotEqual(t, r1.GetMetadata().GetRevision(), r2.GetMetadata().GetRevision())
+
+	// Fetch all resources using paging default.
+	out, nextToken, err = service.ListResources(ctx, 0, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.NotEmpty(t, out)
+	require.Equal(t, []*testResource153{r1, r2}, out)
+
+	// Fetch a paginated list of resources.
+	paginatedOut := make([]*testResource153, 0, 2)
+	numPages := 0
+	for {
+		numPages++
+		out, nextToken, err = service.ListResources(ctx, 1, nextToken)
+		require.NoError(t, err)
+
+		paginatedOut = append(paginatedOut, out...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	require.Equal(t, 2, numPages)
+	require.Equal(t, []*testResource153{r1, r2}, paginatedOut)
+
+	// Fetch a specific service provider.
+	r, err := service.GetResource(ctx, r2.GetMetadata().GetName())
+	require.NoError(t, err)
+	require.Equal(t, r2, r)
+
+	// Try to fetch a resource that doesn't exist.
+	_, err = service.GetResource(ctx, "doesnotexist")
+	require.ErrorIs(t, err, trace.NotFound(`generic resource "doesnotexist" doesn't exist`))
+
+	// Try to create the same resource.
+	_, err = service.CreateResource(ctx, r1)
+	require.ErrorIs(t, err, trace.AlreadyExists(`generic resource "r1" already exists`))
+
+	// Update a resource.
+	r1.Metadata.Labels = map[string]string{"newlabel": "newvalue"}
+	r1, err = service.UpdateResource(ctx, r1)
+	require.NoError(t, err)
+	r, err = service.GetResource(ctx, r1.GetMetadata().GetName())
+	require.NoError(t, err)
+	//nolint:staticcheck // SA1019. Deprecated, but still needed.
+	r.Metadata.Id = r1.Metadata.Id
+	require.Equal(t, r1, r)
+
+	// Update a resource that doesn't exist.
+	doesNotExist := newTestResource153("doesnotexist")
+	_, err = service.UpdateResource(ctx, doesNotExist)
+	require.ErrorIs(t, err, trace.NotFound(`generic resource "doesnotexist" doesn't exist`))
+
+	// Delete a resource.
+	err = service.DeleteResource(ctx, r1.GetMetadata().GetName())
+	require.NoError(t, err)
+	out, nextToken, err = service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Equal(t, []*testResource153{r2}, out)
+
+	// Upsert a resource (create).
+	r1, err = service.UpsertResource(ctx, r1)
+	require.NoError(t, err)
+	out, nextToken, err = service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	//nolint:staticcheck // SA1019. Deprecated, but still needed.
+	out[0].Metadata.Id = r1.Metadata.Id
```

Comment:
This is making a strong ordering assumption, so same comments as above.

Ditto below.